File: C++/2-Medium/00134._加油站/00134._加油站.cpp
#include <iostream>
#include <vector>
using namespace std;

// 函数声明
int canCompleteCircuit(vector<int>& gas, vector<int>& cost);

int main() {
    // 示例输入
    vector<int> gas = {1,2,3,4,5};
    vector<int> cost = {3,4,5,1,2};

    // 调用函数计算结果
    int start = canCompleteCircuit(gas, cost);

    // 输出结果
    cout << "Start from station: " << start << endl;

    return 0;
}

// 函数定义
int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
    int totalTank = 0, currTank = 0, startingStation = 0;
    for (int i = 0; i < gas.size(); i++) {
        int netGain = gas[i] - cost[i];
        totalTank += netGain;
        currTank += netGain;
        // 如果当前油量小于0，重置起始加油站和当前油量
        if (currTank < 0) {
            startingStation = i + 1;
            currTank = 0;
        }
    }
    // 如果总油量大于等于0，返回起始加油站；否则，返回-1
    return totalTank >= 0 ? startingStation : -1;
}

File: Python/2-Medium/00134._加油站/00134._加油站.py
from typing import List

def canCompleteCircuit(gas: List[int], cost: List[int]) -> int:
    """
    计算能否绕环路行驶一周，如果可以，返回起始加油站的索引，否则返回-1。
    
    Args:
    gas: 一个整数列表，表示每个加油站有的汽油量。
    cost: 一个整数列表，表示从当前加油站到下一个加油站的耗油量。
    
    Returns:
    int: 能够绕环路行驶一周的起始加油站索引；如果不存在这样的加油站，则返回-1。
    
    示例:
    >>> canCompleteCircuit([1,2,3,4,5], [3,4,5,1,2])
    3
    >>> canCompleteCircuit([2,3,4], [3,4,3])
    -1
    """
    total_tank, curr_tank, start_station = 0, 0, 0
    
    for i in range(len(gas)):
        total_tank += gas[i] - cost[i]
        curr_tank += gas[i] - cost[i]
        
        # 如果当前累计的油量小于0，意味着不能从start_station到达当前站点的下一个站点
        # 需要将起始点设为当前站点的下一个站点，并重置当前油箱的油量
        if curr_tank < 0:
            start_station = i + 1
            curr_tank = 0
            
    # 如果总油量大于等于总耗油量，则存在至少一个解
    return start_station if total_tank >= 0 else -1

# 测试代码
test_cases = [([1,2,3,4,5], [3,4,5,1,2]), ([2,3,4], [3,4,3])]
results = [canCompleteCircuit(gas, cost) for gas, cost in test_cases]
results